### ISSUE TITLE
About the loss part

### ISSUE THREAD

**Issue by door5719** [andy-yun/pytorch-0.4-yolov3#36]
nPP is larger then nGT a lot , how to solve it case？

**Comment by andy-yun** [comment]
nPP is the candidate of prediction. It is controlled by the confidence value. In code, I use 0.25.
If you increase the value, then you get the small one. But, in the beginning of training, too many zeros are shown. Then you may wonder the training is correct.

**Comment by door5719** [comment]
Thank you sir. I want ask you some question about running. My cpu is Intel Xeon E5-2620 v4 and my gpu is 1080ti x 4. Maybe , function: built_target(running in cpu ) is too slow....T.T, Do you know how to improve efficiency?

**Comment by andy-yun** [comment]
You can modify the operations on CPU or GPU. By comparing the actual execution time, you should modify the source codes. In my opinion, loop computation takes much time on GPU. Thus, I and marvis adopted CPU execution. But, I am not sure, my assumption may not be worked on your environment. There is no unique answer.

**Comment by door5719** [comment]
Thank you sir. I  run  built_target function on GPU but  it still slow. Maybe other better CPU    will run faster than Intel Xeon E5-2620.

**Comment by door5719** [comment]
Sir,  I use my training weights with this repo, I find that : when the objects are large or their distance is big, it can get  better perfomence; if the  objects are small and they are very close to each other ,bad results are got ,such as getting a lot of bboxes with object and they are not precision. What's the reason? I  train the .weights on coco dataset  about 80 epoches.

**Comment by andy-yun** [comment]
That is the weakness of yolo and its derivatives (yolo9000, yolov3). If you want accurate small object detection, I would like to recommend the other methods.

**Comment by door5719** [comment]
Thank you sir.

**Comment by door5719** [comment]
Sir, I also find that train .weights with this repo, confidence seems lager then offical .weights , how to reduce the confidence part of .weights during training process.   Set yolo_layer.py ->class yololayer ()   ->self.rescore = 1 may solve this situation ?

**Comment by andy-yun** [comment]
@door5719 I changed MSELoss and CrossEntropy to BCELoss. BCELoss is greater than MSELoss and CrossEntropy in average.  If you want to small confidence score, you could back to MSELoss and CrossEntropy.

**Comment by door5719** [comment]
Thank you sir，I  will have a try.

**Comment by door5719** [comment]
Sir，something puzzled me when I reading yolo_layer.py. Line :85 gt_box = torch.FloatTensor([gx, gy, gw, gh]) ,gx and gy are  center in feature map( such as 13x13/26x26/52x52 size ) ,gw and  gh are size in input part(such as 416x416). In line 87 :iou = bbox_iou(gt_box, pred_box, x1y1x2y2=False)  ,  go to utils.py , Line 20:def bbox_iou(box1, box2, x1y1x2y2=True): 
in this fuction,coords of box1 and box2 such as  box1[0], box1[1],box1[2], box1[3] should in the same feature map , their size shoud be uniform for the same condition . Maybe I get a wrong comprehension .

**Comment by andy-yun** [comment]
if x1y1x2y2 is True, then (x0, y0, x1, y1) is entered, else (x0, y0, w, h) is entered. 
in bbox_iou, w,h are calculated along to inputs.
